File: SortWare/SortDirectory.vb
Imports System.Xml

Public Class SortDirectory

    Private dir As IO.DirectoryInfo
    Private dirString As String = ""
    Private indent As Integer
    Private tags As String()
    Private isSub As Boolean = False
    Private subDirs As List(Of SortDirectory)
    Private scriptPath As String = ""
    Private convTitle As String = ""

    Private _logReader As IO.StreamReader
    Private _logWriter As IO.StreamWriter

    Private Const TAGFILENAME = "\_tags.txt"
    Public ReadOnly Property type As SortSettings.dirType

    Public Sub New(ByVal _dir As String, Optional ByVal _in As Integer = 0)
        type = SortSettings.dirType.ERRORDIR
        dirString = _dir
        dir = New IO.DirectoryInfo(_dir)

        If _in > 0 Then
            indent = _in
        Else
            indent = 0
        End If
    End Sub

    Public Sub New(ByVal _dir As String, ByVal _in As Integer, ByVal _type As SortSettings.dirType, ByVal Optional isSubDir As Boolean = False, ByVal Optional name As String = "", ByVal Optional script As String = "", ByVal Optional convType As String = "")
        type = _type
        dirString = _dir
        dir = New IO.DirectoryInfo(_dir)

        If _in > 0 Then
            indent = _in
        Else
            indent = 0
        End If

        isSub = isSubDir

        scriptPath = script
        convTitle = name

        If Me.hasTags() Then
            tags = Me.getTagsFromFile()
        End If
        If Me.hasSubs() Then
            Me.loadSubs()
        End If

    End Sub

    Public Function exists() As Boolean
        Return dir.Exists And IO.Directory.Exists(dir.FullName)
    End Function

    Public Sub addSubDir(ByVal s As String)
        If type = SortSettings.dirType.MAINDIR Then
            If s.Contains(Me.fullName) Then
                If subDirs Is Nothing Then subDirs = New List(Of SortDirectory)
            End If
            subDirs.Add(New SortDirectory(s, Me.indent + 1, SortSettings.dirType.MAINDIR, True))
        End If
    End Sub

    Public Sub addSubDir(ByVal sd As SortDirectory)
        If type = SortSettings.dirType.MAINDIR Then
            If sd.fullName.Contains(Me.fullName) Then
                If subDirs Is Nothing Then subDirs = New List(Of SortDirectory)
            End If
            subDirs.Add(sd)
        End If
    End Sub

    Public Sub loadSubs()
        Dim xdoc As XmlDocument = New XmlDocument
        xdoc.Load(fullName() & "\subSortSettings.xml")

        Dim smain = xdoc.GetElementsByTagName(SortSettings.XMLNODESUBS)

        For Each s As XmlNode In smain
            For Each c As XmlNode In s.ChildNodes
                If c.Name = "dir" Then
                    Me.addSubDir(c.FirstChild.Value.Trim)
                End If
            Next
        Next

    End Sub
    Public Sub saveSubs()
        If Me.hasSubs Then

            Dim xws As XmlWriterSettings = New XmlWriterSettings
            xws.Indent = True

            Using writer As XmlWriter = XmlWriter.Create(fullName() + "\subSortSettings.xml", xws)
                writer.WriteStartDocument()
                writer.WriteStartElement("rootDir")
                writer.WriteAttributeString("dir", fullName)

                If subDirs Is Nothing Then
                    writer.WriteStartElement("subs")
                    For Each m In subDirs
                        writer.WriteStartElement("dir")
                        If m.hasSubs() Then
                            writer.WriteAttributeString("hasSub", "true")
                            m.saveSubs()
                        End If
                        writer.WriteString(m.fullName)
                        writer.WriteFullEndElement()
                    Next
                    writer.WriteFullEndElement()
                End If

                writer.WriteFullEndElement()
                writer.Close()
            End Using




            'Using _sortSettingsWriter = New IO.StreamWriter(Me.fullName + "\.SubSortSettings.txt")
            '    _sortSettingsWriter.Write(New SubSortSettings(Me, subDirs).toString)
            'End Using
            'For Each sd In getSubs()
            '    sd.saveSubs()
            'Next
        End If
    End Sub


    Public Function hasSubs() As Boolean
        Return subDirs IsNot Nothing Or (Me.exists AndAlso IO.File.Exists(Me.fullName + "\subSortSettings.xml"))
    End Function

    Public Function getSubs() As List(Of SortDirectory)
        Return subDirs
    End Function

    Public Function hasTags() As Boolean
        Return Me.exists AndAlso IO.File.Exists(Me.fullName & TAGFILENAME)
    End Function

    Public Function getTags() As String()
        Return tags
    End Function

    Private Function getTagsFromFile() As String()
        If Not Me.hasTags Then
            Return Nothing
        End If
        Dim ret As New List(Of String)
        Dim reader As IO.TextReader = New IO.StreamReader(fullName() & TAGFILENAME)
        Dim fullString = Split(reader.ReadToEnd.Trim, vbNewLine)
        reader.Close()
        Return fullString
    End Function

    Public Sub saveTags(ByVal _tags As String)
        'If Me.hasTags Then
        Dim writer = New System.IO.StreamWriter(fullName() & TAGFILENAME, False)
        writer.Write(_tags.Trim)
        writer.Close()
        'End If
        tags = Me.getTagsFromFile()
    End Sub

    Public Function fullName() As String
        Return dir.FullName
    End Function

    Public Function getParent() As SortDirectory
        Return New SortDirectory(IO.Directory.GetParent(dir.FullName).FullName, indent - 1)
    End Function

    Public Function getName() As String
        Return dir.Name
    End Function

    Public Function getConvTitle() As String
        Return If(type = SortSettings.dirType.CONVERTDIR, convTitle, "")
    End Function

    Public Function isSubDir() As Boolean
        Return isSub
    End Function

    Public Function getScriptPath() As String
        Return scriptPath
    End Function

    Public Function matchesFilter(ByVal filter As String) As Boolean
        Dim ret As Boolean = False

        If Not String.IsNullOrWhiteSpace(filter) Then
            If getName.ToLower.Contains(filter.ToLower) Then
                ret = True
            End If

            If hasSubs() Then
                For Each d In getSubs()
                    If d.matchesFilter(filter) Then
                        ret = True
                    End If
                Next
            End If
        End If

        Return ret
    End Function

    Public Sub SetDir(ByVal newDir As String)
        dir = New IO.DirectoryInfo(newDir)
    End Sub

    Public Function isValid() As Boolean
        Dim ret As Boolean = True

        If type = SortSettings.dirType.CONVERTDIR Then
            ret = ret And Not (scriptPath = Nothing OrElse Not IO.Directory.Exists(scriptPath))
        End If

        ret = ret And Me.exists

        Return ret
    End Function

    Public Overrides Function ToString() As String
        Return dir.Name.PadLeft((indent * 3) + dir.Name.Length, Convert.ToChar(" "))
    End Function

End Class

File: SortWare/SortDirectory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Xml;
using System.Xml.Serialization;
using Microsoft.VisualBasic;

namespace SortWare
{

  public class SortDirectory
  {

    private System.IO.DirectoryInfo m_cDir;
    private string dirString = "";
    private int indent;
    private List<DirTag> m_cTags = new List<DirTag>();
    private bool isSub = false;
    private List<SortDirectory> m_cSubDirs;
    private string scriptPath = "";
    private string convTitle = "";

    private System.IO.StreamReader _logReader;
    private System.IO.StreamWriter _logWriter;

    private const string TAGFILENAME = @"\_tags.txt";

    [XmlAttribute("dir_type")]
    public SortSettings.dirType type { get; set; }

    [XmlArray("sub_dirs")]
    [XmlArrayItem("sub-dir")]
    public List<SortDirectory> SubDirs { get { return m_cSubDirs; } set { m_cSubDirs = value; } }

    [XmlAttribute("path")]
    public string fullName { get { return m_cDir.FullName; } set { m_cDir = new System.IO.DirectoryInfo(value); } }

    [XmlArray("tags")]
    [XmlArrayItem("tag")]
    public List<DirTag> DirTags { get { return m_cTags; } set { m_cTags = value; } }

    public SortDirectory() { }

    public SortDirectory(string _dir, int _in = 0)
    {
      @type = SortSettings.dirType.ERRORDIR;
      dirString = _dir;
      m_cDir = new System.IO.DirectoryInfo(_dir);

      if (_in > 0)
      {
        indent = _in;
      }
      else
      {
        indent = 0;
      }
    }

    public SortDirectory(string _dir, int _in, SortSettings.dirType _type, bool isSubDir = false, string name = "", string script = "", string convType = "")
    {
      @type = _type;
      dirString = _dir;
      m_cDir = new System.IO.DirectoryInfo(_dir);

      if (_in > 0)
      {
        indent = _in;
      }
      else
      {
        indent = 0;
      }

      isSub = isSubDir;

      scriptPath = script;
      convTitle = name;


      if (hasSubs())
      {
        loadSubs();
      }

    }

    public void AddTag(DirTag tag)
    {
      DirTags.Add(tag);
      DirTags.Sort((x,y) => x.Key.CompareTo(y.Key));
    }

    public void RemoveTag(DirTag tag)
    {
      DirTags.Remove(tag);
      DirTags.Sort((x, y) => x.Key.CompareTo(y.Key));
    }

    public bool exists()
    {
      return m_cDir.Exists & System.IO.Directory.Exists(m_cDir.FullName);
    }

    public void addSubDir(string s)
    {
      if (@type == SortSettings.dirType.MAINDIR)
      {
        if (s.Contains(fullName))
        {
          if (m_cSubDirs is null)
            m_cSubDirs = new List<SortDirectory>();
        }
        m_cSubDirs.Add(new SortDirectory(s, indent + 1, SortSettings.dirType.MAINDIR, true));
      }
    }

    public void addSubDir(SortDirectory sd)
    {
      if (@type == SortSettings.dirType.MAINDIR)
      {
        if (sd.fullName.Contains(fullName))
        {
          if (m_cSubDirs is null)
            m_cSubDirs = new List<SortDirectory>();
        }
        m_cSubDirs.Add(sd);
      }
    }

    public void loadSubs()
    {
      var xdoc = new XmlDocument();
      xdoc.Load(fullName + @"\subSortSettings.xml");

      var smain = xdoc.GetElementsByTagName(SortSettings.XMLNODESUBS);

      foreach (XmlNode s in smain)
      {
        foreach (XmlNode c in s.ChildNodes)
        {
          if (c.Name == "dir")
          {
            addSubDir(c.FirstChild.Value.Trim());
          }
        }
      }

    }
    public void saveSubs()
    {
      if (hasSubs())
      {

        var xws = new XmlWriterSettings();
        xws.Indent = true;

        using (var writer = XmlWriter.Create(fullName + @"\subSortSettings.xml", xws))
        {
          writer.WriteStartDocument();
          writer.WriteStartElement("rootDir");
          writer.WriteAttributeString("dir", fullName);

          if (m_cSubDirs is null)
          {
            writer.WriteStartElement("subs");
            foreach (var m in m_cSubDirs)
            {
              writer.WriteStartElement("dir");
              if (m.hasSubs())
              {
                writer.WriteAttributeString("hasSub", "true");
                m.saveSubs();
              }
              writer.WriteString(m.fullName);
              writer.WriteFullEndElement();
            }
            writer.WriteFullEndElement();
          }

          writer.WriteFullEndElement();
          writer.Close();
        }




        // Using _sortSettingsWriter = New IO.StreamWriter(Me.fullName + "\.SubSortSettings.txt")
        // _sortSettingsWriter.Write(New SubSortSettings(Me, subDirs).toString)
        // End Using
        // For Each sd In getSubs()
        // sd.saveSubs()
        // Next
      }
    }


    public bool hasSubs() => m_cSubDirs is not null && m_cSubDirs.Count > 0;

    public ref List<SortDirectory> getSubs() => ref m_cSubDirs;

    public bool hasTags() => exists() && System.IO.File.Exists(fullName + TAGFILENAME);


    public SortDirectory getParent()
    {
      return new SortDirectory(System.IO.Directory.GetParent(m_cDir.FullName).FullName, indent - 1);
    }

    public string getName()
    {
      return m_cDir.Name;
    }

    public string getConvTitle()
    {
      return @type == SortSettings.dirType.CONVERTDIR ? convTitle : "";
    }

    public bool isSubDir()
    {
      return isSub;
    }

    public string getScriptPath()
    {
      return scriptPath;
    }

    public bool matchesFilter(string filter)
    {
      bool ret = false;

      if (!string.IsNullOrWhiteSpace(filter))
      {
        if (getName().ToLower().Contains(filter.ToLower()))
        {
          ret = true;
        }

        if (hasSubs())
        {
          foreach (var d in getSubs())
          {
            if (d.matchesFilter(filter))
            {
              ret = true;
            }
          }
        }
      }

      return ret;
    }

    public void SetDir(string newDir)
    {
      m_cDir = new System.IO.DirectoryInfo(newDir);
    }

    public bool isValid()
    {
      bool ret = true;

      if (@type == SortSettings.dirType.CONVERTDIR)
      {
        ret = ret & !(string.IsNullOrEmpty(scriptPath) || !System.IO.Directory.Exists(scriptPath));
      }

      ret = ret & exists();

      return ret;
    }

    public override string ToString()
    {
      return m_cDir.Name.PadLeft(indent * 3 + m_cDir.Name.Length, Convert.ToChar(" "));
    }

  }
}